### ISSUE TITLE
stateless boot node looses network connection at dhcp lease expiration time.

### ISSUE THREAD

**Issue by ralphbellofatto** [xcat2/xcat-core#1972]
We noticed this first that when we booted a diskless image, that it would loose its network connectivity after 12 hours.

After some further isolation we determined that this 12 hours correlated to the dhcp leases time.

We followed the directions to define a node:
http://xcat-docs.readthedocs.io/en/stable/guides/admin-guides/manage_clusters/ppc64le/discovery/manually_define.html

and ended up with the following lsdef clause...

```
c460c037:
    objtype=node
    arch=ppc64le
    bmc=10.2.3.1
    bmcport=0
    bmcvlantag=2
    chain=runcmd=bmcsetup,standby
    cons=ipmi
    consoleondemand=0
    cpucount=160
    cputype=POWER8NVL (raw), altivec supported
    disksize=sda:2000GB,sdb:2000GB
    groups=all,compute,c460f03
    hostnames=c460f03u03
    initrd=xcat/osimage/rh72_op5_compute/initrd-stateless.gz
    installnic=mac
    ip=10.1.3.1
    kcmdline=imgurl=http://!myipfn!:80//install/netboot/rhels7.2/ppc64le/rh72_op5_compute/rootimg.cpio.gz XCAT=!myipfn!:3001 NODE=c460c037 FC=0 BOOTIF=70:e2:84:14:0d:e6 
    kernel=xcat/osimage/rh72_op5_compute/kernel
    mac=70:e2:84:14:0d:e6
    memory=523125MB
    mgt=ipmi
    mtm=8335-GTB
    netboot=petitboot
    nodetype=mp
    os=rhels7.2
    postbootscripts=confignics,setroute,setupntp
    primarynic=mac
    profile=compute
    provmethod=rh72_op5_compute
    rack=c460f03
    serial=100A49A
    supportedarchs=ppc64
    switch=c460tors03
    switchport=3
    unit=u03
```

To see the problem without waiting 12 hours and to isolate parts of our software loadout we pointed the node at the default netboot-compute image we got with copycds and then reduced the dhcpleases time to 5 minutes with the following procedure:

```
genimage rhels7.2-ppc64le-netboot-compute
packimage rhels7.2-ppc64le-netboot-compute
nodeset c460c037 osimage=rhels7.2-ppc64le-netboot-compute


# keep the leases time short to stimulate the problem.
chdef -t site clustersite dhcplease=300
makedhcp -n
makedhcp -a
systemctl restart dhcpd

# reboot the node.
rpower c460c037 off;
rpower c460c037 on;

```

verify the lease is a short one:

```
    xdsh c460c037 cat /var/lib/dhclient/dhclient-enP5p7s0f0.leases
    lease {
      interface "enP5p7s0f0";
      fixed-address 10.1.3.1;
      filename "/yaboot";
      option subnet-mask 255.255.0.0;
      option routers 10.1.0.1;
      option dhcp-lease-time 300;
      option dhcp-message-type 5;
      option domain-name-servers 10.1.0.1;
      option dhcp-server-identifier 10.1.0.1;
      option domain-search "c460cluster.net.";
      option host-name "c460c037";
      option domain-name "c460cluster.net";
      renew 5 2016/10/14 21:39:39;
      rebind 5 2016/10/14 21:42:03;
      expire 5 2016/10/14 21:42:41;
    }
```

Then invoke a loop to on the system console to monitor the lease expiration...

```
rcons c460c037
# login.

while true; do
ip a show dev enP5p7s0f0;
sleep 30;
done
```

And we notice:

```
[Fri Oct 14 17:39:17 2016][root@c460c037 ~]# while true; do
[Fri Oct 14 17:39:17 2016]> ip a show dev enP5p7s0f0;
[Fri Oct 14 17:39:17 2016]> sleep 30;
[Fri Oct 14 17:39:17 2016]> done
[Fri Oct 14 17:39:17 2016]2: enP5p7s0f0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP qlen 1000
[Fri Oct 14 17:39:17 2016]    link/ether 70:e2:84:14:0d:e6 brd ff:ff:ff:ff:ff:ff
[Fri Oct 14 17:39:17 2016]    inet 10.1.3.1/16 brd 10.1.255.255 scope global dynamic enP5p7s0f0
[Fri Oct 14 17:39:17 2016]       valid_lft 207sec preferred_lft 207sec
[Fri Oct 14 17:39:17 2016]    inet6 fe80::72e2:84ff:fe14:de6/64 scope link
[Fri Oct 14 17:39:17 2016]       valid_lft forever preferred_lft forever
[Fri Oct 14 17:39:47 2016]2: enP5p7s0f0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP qlen 1000
[Fri Oct 14 17:39:47 2016]    link/ether 70:e2:84:14:0d:e6 brd ff:ff:ff:ff:ff:ff
[Fri Oct 14 17:39:47 2016]    inet 10.1.3.1/16 brd 10.1.255.255 scope global dynamic enP5p7s0f0
[Fri Oct 14 17:39:47 2016]       valid_lft 177sec preferred_lft 177sec
[Fri Oct 14 17:39:47 2016]    inet6 fe80::72e2:84ff:fe14:de6/64 scope link
[Fri Oct 14 17:39:47 2016]       valid_lft forever preferred_lft forever
...
[Fri Oct 14 17:42:17 2016]    link/ether 70:e2:84:14:0d:e6 brd ff:ff:ff:ff:ff:ff
[Fri Oct 14 17:42:17 2016]    inet 10.1.3.1/16 brd 10.1.255.255 scope global dynamic enP5p7s0f0
[Fri Oct 14 17:42:17 2016]       valid_lft 27sec preferred_lft 27sec
[Fri Oct 14 17:42:17 2016]    inet6 fe80::72e2:84ff:fe14:de6/64 scope link
[Fri Oct 14 17:42:17 2016]       valid_lft forever preferred_lft forever

[Fri Oct 14 17:42:47 2016]2: enP5p7s0f0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP qlen 1000
[Fri Oct 14 17:42:47 2016]    link/ether 70:e2:84:14:0d:e6 brd ff:ff:ff:ff:ff:ff
[Fri Oct 14 17:42:47 2016]    inet6 fe80::72e2:84ff:fe14:de6/64 scope link
[Fri Oct 14 17:42:47 2016]       valid_lft forever preferred_lft forever
[Fri Oct 14 17:43:17 2016]2: enP5p7s0f0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP qlen 1000
[Fri Oct 14 17:43:17 2016]    link/ether 70:e2:84:14:0d:e6 brd ff:ff:ff:ff:ff:ff
[Fri Oct 14 17:43:17 2016]    inet6 fe80::72e2:84ff:fe14:de6/64 scope link
[Fri Oct 14 17:43:17 2016]       valid_lft forever preferred_lft forever
```

notice that the  the inet clause is no longer there, this port no longer has an ip address...

and from the management node the node is no longer reachable.

```
[root@c460mgt01 tmp]# xdsh c460c037 who
c460c037: ssh: connect to host c460c037 port 22: No route to host
```

**Comment by immarvin** [comment]
seems to be a similar  issue with https://github.com/xcat2/xcat-core/issues/942, 

hi @ralphbellofatto , would you please check the following 2 points on the diskless node?
1.  whether "NetworkManager" is running with ``systemctl status NetworkManager.service`` 
2. whether dhclient process running with ``ps -aux|grep dhclient``

thanks

**Comment by ralphbellofatto** [comment]
[root@c460c037 ~]# systemctl status NetworkManager.service
â— NetworkManager.service
   Loaded: not-found (Reason: No such file or directory)
   Active: inactive (dead)
[root@c460c037 ~]# ps -aux|grep dhclient
root      10840  0.0  0.0   3840  1472 hvc0     S+   13:24   0:00 grep --color=auto dhclient

**Comment by immarvin** [comment]
This problem is caused by the fact that NetworkManager is not in pkglist of the rh7 diskless osimage, hence there is no "dhclient" process running on the live diskless node.  
fixed in https://github.com/xcat2/xcat-core/pull/1977

with this fix, the NetworkManager service will be enabled on rh7 diskless node boot, this might introduce some problem for some xCAT postscripts, see https://github.com/xcat2/xcat-core/issues/104 for details.

**Comment by ralphbellofatto** [comment]
We applied the change of adding the NetworkManager to our configuraiton.

While we no longer loose our IP address (which is good), I don't think that it is quite working right either:

Set up the experiment again to see if this fixes the problem:
```
chdef -t site clustersite dhcplease=600
makedhcp -n
makedhcp -a
systemctl restart dhcp

rpower c460c037 off
rpower c460c037 on

rcons c460c037
# login.

while true; do
ip a show dev enP5p7s0f0;
sleep 30;
done

```

And we stay up.  But...

We show the network manager up.
```
[root@c460c037 ~]# systemctl status NetworkManager
â— NetworkManager.service - Network Manager
   Loaded: loaded (/usr/lib/systemd/system/NetworkManager.service; enabled; vendor preset: enabled)
   Active: active (running) since Mon 2016-10-17 09:02:48 EDT; 17min ago
 Main PID: 2014 (NetworkManager)
   CGroup: /system.slice/NetworkManager.service
           â”œâ”€2014 /usr/sbin/NetworkManager --no-daemon
           â””â”€2509 /sbin/dhclient -d -q -sf /usr/libexec/nm-dhcp-helper -pf /var/run/dhclient-enP5p7s0f0...

Oct 17 09:16:44 c460c037 NetworkManager[2014]: <info>    address 10.1.3.1
Oct 17 09:16:44 c460c037 NetworkManager[2014]: <info>    plen 16 (255.255.0.0)
Oct 17 09:16:44 c460c037 NetworkManager[2014]: <info>    gateway 10.1.0.1
Oct 17 09:16:44 c460c037 NetworkManager[2014]: <info>    server identifier 10.1.0.1
Oct 17 09:16:44 c460c037 NetworkManager[2014]: <info>    lease time 600
Oct 17 09:16:44 c460c037 NetworkManager[2014]: <info>    hostname 'c460c037'
Oct 17 09:16:44 c460c037 NetworkManager[2014]: <info>    nameserver '10.1.0.1'
Oct 17 09:16:44 c460c037 NetworkManager[2014]: <info>    domain name 'c460cluster.net'
Oct 17 09:16:44 c460c037 NetworkManager[2014]: <info>    domain search 'c460cluster.net.'
Oct 17 09:16:44 c460c037 NetworkManager[2014]: <info>  (enP5p7s0f0): DHCPv4 state changed bound -> bound
[root@c460c037 ~]# ps -A | grep dhclient
  2509 ?        00:00:00 dhclient
```

but, 
```

[root@c460c037 ~]# cat /var/lib/dhclient/dhclient-enP5p7s0f0.leases
lease {
  interface "enP5p7s0f0";
  fixed-address 10.1.3.1;
  filename "/yaboot";
  option subnet-mask 255.255.0.0;
  option routers 10.1.0.1;
  option dhcp-lease-time 600;
  option dhcp-message-type 5;
  option domain-name-servers 10.1.0.1;
  option dhcp-server-identifier 10.1.0.1;
  option domain-search "c460cluster.net.";
  option host-name "c460c037";
  option domain-name "c460cluster.net";
  renew 1 2016/10/17 13:05:03;
  rebind 1 2016/10/17 13:09:13;
  expire 1 2016/10/17 13:10:28;
}
```

We don't loose our network connection, which is goodness, but we don't renew the dhcp lease either. 
Mon Oct 17 09:19:54 EDT 2016

The dhcp lease is 10 minutes long,   and by 9:19 EST (13:19 UTC) it should have renewed the lease.

So, i think something is still a little loose.

**Comment by mattaezell** [comment]
@ralphbellofatto We are currently using both the **hardeths** and **confignics -s** postscripts, although I'm not sure if we actually need both.  We don't think our nodes should depend on DHCP after boot - these are static addresses that won't change.  There's no point sending DHCP packets every 12 hours.  We don't want an extra daemon either, since we are trying to minimize jitter.

**Comment by ralphbellofatto** [comment]
Matt,  Thanks for the information.  

Technically, if you have this in your lsdef, you are using dhcp in your boot up, xcat has setup the dhcp leases file to always hand out the same address.

This stanza should work, but does not work on our research cluster and does work on the EA3.

```
Object name: ray20
    addkcmdline=modprobe.blacklist=nvidiafb
    arch=ppc64le
    bmc=192.168.0.20
    ip=192.168.64.20
    mac=70:e2:84:14:12:d7!*NOIP*|70:e2:84:14:12:d8
```

One potential workaround for this problem is to include the following attribute in the lsdef:
```
nicips.enP5p7s0f1=192.168.64.20
```

This causes confignics to write out that address into:  /etc/sysconfig/network-scripts/ifcfg-enP5p7s0f1,  
When this is done, dhcp hands out the ip=192.168.64.20 during the ipl process until the confignics script is run.  At that point /etc/sysconfig/network-scripts/ifcfg-enP5p7s0f1 is written/rewritten with a static address, and the dhcp lease is no longer renewed.

Doing this masks the original problem.  But since the directions published by xCAT say this stanza should work: 
```
```
Object name: ray20
    addkcmdline=modprobe.blacklist=nvidiafb
    arch=ppc64le
    bmc=192.168.0.20
    ip=192.168.64.20
    mac=70:e2:84:14:12:d7!*NOIP*|70:e2:84:14:12:d8
```
We are trying to isolate this problem, either fix the problem for good, or get the documentation changed with at least an erratta for this circumstance.


We noticed that the EA3 system being tested for CORAL does NOT have the NetworkManager setup but does not exhibit this problem.

We traced this to the existence of this file on the EA3 system, which did not get created by research cluster.
```
cat /etc/sysconfig/network-scripts/ifcfg-enP5p7s0f1 
DEVICE=enP5p7s0f1
BOOTPROTO=dhcp
HWADDR=70:e2:84:14:12:d8
ONBOOT=yes
```

This is the code in: /opt/xcat/share/xcat/netboot/rh/dracut_033/xcatroot that created the file on the ea3 system.
```
if [ -z $STATEMNT ]; then
  for lf in /tmp/dhclient.*.lease; do
    netif=${lf#*.}
    netif=${netif%.*}
    cp $lf  "$NEWROOT/var/lib/dhclient/dhclient-$netif.leases"
    [ "$xcatdebugmode" > "0" ] && logger -t xcat -p debug "saving $NEWROOT/var/lib/dhclient/dhclient-$netif.leases"
  done

  if [ ! -z "$ifname" ]; then
    MACX=${ifname#*:}
    ETHX=${ifname%:$MACX*}
  elif [ ! -z "$netdev" ]; then
    ETHX=$netdev
    MACX=`ip link show $netdev | grep ether | awk '{print $2}'`
  elif [ ! -z "$BOOTIF" ]; then
    MACX=$BOOTIF
    #ETHX=`ifconfig |grep -i $BOOTIF | awk '{print $1}'`
    ETHX=` ip  -oneline link show |grep -i $BOOTIF|awk -F ':' '{print $2}'|grep -o "[^ ]\+\( \+[^ ]\+\)*"`
  fi

  if [ ! -z "$MACX" ] && [ ! -z "$ETHX" ]; then
    if [ ! -e $NEWROOT/etc/sysconfig/network-scripts/ifcfg-$ETHX ]; then
        [ "$xcatdebugmode" > "0" ] && logger -t xcat -p debug "creating $NEWROOT/etc/sysconfig/network-scripts/ifcfg-$ETHX"
        touch $NEWROOT/etc/sysconfig/network-scripts/ifcfg-$ETHX
    fi
    [ "$xcatdebugmode" > "0" ] && logger -t xcat -p debug "writing $NEWROOT/etc/sysconfig/network-scripts/ifcfg-$ETHX: DEVICE=$ETHX;BOOTPROTO=dhcp;HWADDR=$MACX;ONBOOT=yes"
    echo "DEVICE=$ETHX" > $NEWROOT/etc/sysconfig/network-scripts/ifcfg-$ETHX
    echo "BOOTPROTO=dhcp" >> $NEWROOT/etc/sysconfig/network-scripts/ifcfg-$ETHX
    echo "HWADDR=$MACX" >> $NEWROOT/etc/sysconfig/network-scripts/ifcfg-$ETHX
    echo "ONBOOT=yes" >> $NEWROOT/etc/sysconfig/network-scripts/ifcfg-$ETHX
  fi
fi
```

We are doing traces on our research cluster to figure out why this file is not getting written on some systems and is getting written on others.

**Comment by mattaezell** [comment]
Is EA3 using xCAT 2.12.2 and your research cluster using something newer (2.12.3 or a snapshot)?  The diskless initrd switched to using systemd in dracut, which likely changed the way dhcp is being handled.

**Comment by ralphbellofatto** [comment]
Interesting observation:

This is the research cluster:
```
Version 2.12.3 (git commit ee11ecddb7e3c323b11857826a072303ae5a7b20, built Thu Sep 22 06:16:00 EDT 2016)
```
and

this is the EA3 system:
```
[root@ray1 ~]# lsxcatd -v
Version 2.12.2 (git commit e4a7359d31008d81c4c57e0857d5f09009a3269c, built Thu Aug 18 06:43:29 EDT 2016)
```

We are on a snaptshot to pick up some recent changes that allow us at genimage postinstall time to know what osimage we are installing.

It is fortunate that we did this, because there was another change in dracut that some how changed the way wget was displaying progress, turning a 3 minute download of the rootimage dir into a 3 hour download, which we got corrected by reporting it to the xCAT team and the code to fix it, (which was adding the -q, parameter to the xcatroot wget function.

**Comment by ralphbellofatto** [comment]
With this change in dracut, it would look like the systems configured as recommended in the xCAT documentaiton will exhibit the problem we experienced here, without some sort of workaround.

This would indicate that either something needs to compensate for this in the confignics scripts, the xcatroot script (hooked into dracut) or the documentation mentioning configuring the system like this:

http://xcat-docs.readthedocs.io/en/stable/guides/admin-guides/manage_clusters/ppc64le/discovery/manually_define.html

**Comment by ralphbellofatto** [comment]
My working theory, which i have not been able to verify, because EA3 is being activly being used for testing. is that the dracut scripts that invoke the xcatroot script, or are invoked by xcatroot set one of the following variables on EA3 (xcat version 2.12.2).

$ifname, $netdev, or $BOOTIF.

NOTEP: $BOOTIF is on the command-line, so i would suspect this is the variable that was, and is no longer getting set in xcat 2.12.3.

**Comment by mattaezell** [comment]
@ralphbellofatto Can you try the patch from #1980 ?

**Comment by ralphbellofatto** [comment]
Thanks, I'll give that a spin.

**Comment by ralphbellofatto** [comment]
adding the above patch referenced by matt yieled the expected file:
```
[root@c460c037 ~]# cat /etc/sysconfig/network-scripts/ifcfg-enP5p7s0f0
DEVICE=enP5p7s0f0
BOOTPROTO=dhcp
HWADDR=70:e2:84:14:0d:e6
ONBOOT=yes
```

Now i need to reset up the experiment with the shortened lease time and remove the NetworkManager to see if we don't loose the ip address again.

**Comment by ralphbellofatto** [comment]
After setting up the experiment with the shortened lease time and removing the "NetworkManager" we have verified that having this file causes the leases to be properly renewed as evidenced by:
```
[root@c460c037 ~]# cat /var/lib/dhclient/dhclient-enP5p7s0f0.leases
lease {
  interface "enP5p7s0f0";
  fixed-address 10.1.3.1;
  filename "/yaboot";
  option subnet-mask 255.255.0.0;
  option dhcp-lease-time 600;
  option routers 10.1.0.1;
  option dhcp-message-type 5;
  option dhcp-server-identifier 10.1.0.1;
  option domain-name-servers 10.1.0.1;
  option domain-search "c460cluster.net.";
  option host-name "c460c037";
  option domain-name "c460cluster.net";
  renew 2 2016/10/18 15:53:06;
  rebind 2 2016/10/18 15:57:58;
  expire 2 2016/10/18 15:59:13;
}
lease {
  interface "enP5p7s0f0";
  fixed-address 10.1.3.1;
  filename "/yaboot";
  option subnet-mask 255.255.0.0;
  option routers 10.1.0.1;
  option dhcp-lease-time 600;
  option dhcp-message-type 5;
  option domain-name-servers 10.1.0.1;
  option dhcp-server-identifier 10.1.0.1;
  option domain-search "c460cluster.net.";
  option ntp-servers 10.1.0.1;
  option host-name "c460c037";
  option domain-name "c460cluster.net";
  renew 2 2016/10/18 15:56:17;
  rebind 2 2016/10/18 16:00:21;
  expire 2 2016/10/18 16:01:36;
}
lease {
  interface "enP5p7s0f0";
  fixed-address 10.1.3.1;
  filename "/yaboot";
  option subnet-mask 255.255.0.0;
  option routers 10.1.0.1;
  option dhcp-lease-time 600;
  option dhcp-message-type 5;
  option domain-name-servers 10.1.0.1;
  option dhcp-server-identifier 10.1.0.1;
  option domain-search "c460cluster.net.";
  option ntp-servers 10.1.0.1;
  option host-name "c460c037";
  option domain-name "c460cluster.net";
  renew 2 2016/10/18 16:01:07;
  rebind 2 2016/10/18 16:05:04;
  expire 2 2016/10/18 16:06:19;
}
```

**Comment by ralphbellofatto** [comment]
Victor and I were trying to figure out why dracut with version 2.12.2 used one method in its init scripts, and xcat 2.12.3 used systemd.

I noticed the following difference between 2.12.2 and the HEAD of the master branch.
[Report.txt](https://github.com/xcat2/xcat-core/files/537360/Report.txt)

[diff-2.12.2-to-now.html.zip](https://github.com/xcat2/xcat-core/files/537361/diff-2.12.2-to-now.html.zip)


I suspect that including the systemd into dracut on a 7.2 installation caused it to favor using it.

**Comment by mattaezell** [comment]
Yes, I believe commit c0194840 changed it

**Comment by immarvin** [comment]
hi @mattaezell @ralphbellofatto , thanks for your investigation. 
"systemd" dracut module was introduced on kdump support, since the "kdump" dracut module depends on it.  We also noticed that after "systemd" module is included in initrd, the variables in the kernel cmdline are not set automatically, we added some code to parse and set some variables but forgot to the  "ifname", "netdev" and "BOOTIF".
